src/NHibernate/Loader/DynamicBatchingHelper.cs

```diff
@@ -0,0 +1,58 @@
+using System;
+using NHibernate.Criterion;
+using NHibernate.Engine;
+using NHibernate.SqlCommand;
+
+namespace NHibernate.Loader
+{
+	internal class DynamicBatchingHelper
+	{
+		private static string BatchIdPlaceholder = "$$BATCH_ID_PLACEHOLDER$$";
+
+		public static SqlStringBuilder BuildBatchFetchRestrictionFragment()
+		{
+			return new SqlStringBuilder(1).Add(DynamicBatchingHelper.BatchIdPlaceholder);
+		}
+
+		public static void ExpandBatchIdPlaceholder(SqlString sqlString, QueryParameters queryParameters, string[] columns, Dialect.Dialect dialect, out Parameter[] parameters, out SqlString result)
+		{
+			var bogusParam = Parameter.Placeholder;
+			var wherePart = InExpression.GetSqlString(columns, queryParameters.PositionalParameterValues.Length, dialect, bogusParam);
+			var paramsCount = wherePart.GetParameterCount();
+			parameters = new Parameter[paramsCount];
+			for (var i = 0; i < parameters.Length; i++)
+			{
+				parameters[i] = Parameter.Placeholder;
+			}
+
+			wherePart.SubstituteBogusParameters(parameters, bogusParam);
+			result = sqlString.ReplaceLast(DynamicBatchingHelper.BatchIdPlaceholder, wherePart);
+		}
+
```

Comment:
Please split into 2 methods. One to build where statement and another one to replace placeholder with that where statement.

Reply:
Ok. But common it's 10 lines method - IMO splitting doesn't improve anything.

Reply:
This method just makes zero sense.

Reply:
Meaning? It's reused in multiple places

Reply:
> Meaning?

1. The method does not have a clear purpose it does a lot of half things.
2. It has 2 unrelated `out` parameters.

> It's reused in multiple places

Here are the 2 places that user that method:

```csharp
DynamicBatchingHelper.ExpandBatchIdPlaceholder(sqlString, queryParameters, columns, Factory.Dialect, out var parameters, out var result);
parameterSpecifications.UnionWith(CreateParameterSpecificationsAndAssignBackTrack(parameters));
```

```csharp
DynamicBatchingHelper.ExpandBatchIdPlaceholder(sqlString, queryParameters, columns, Factory.Dialect, out var parameters, out var result);
parameterSpecifications.UnionWith(CreateParameterSpecificationsAndAssignBackTrack(parameters));
```

Spot the difference... 

`CreateParameterSpecificationsAndAssignBackTrack` are also identical with only difference is the key type.

Reply:
Purpose - minimize code duplication.

It returns SQL command and parameters used in this command - looks pretty related to me.

`CreateParameterSpecificationsAndAssignBackTrack` - same name, different methods  . I don't want to bother unifying it.

Reply:
> `CreateParameterSpecificationsAndAssignBackTrack` - same name, different methods. 

It is exactly same method, just differs in one thing that could be an argument. 

But after a close inspection turns out that the whole construct is just a cart in front of a horse.